File: 152. Maximum Product Subarray/dp-kadane.java
class Solution {
    public int maxProduct(int[] nums) {
        
        int minSoFar = nums[0], maxSoFar = nums[0];
        int max = nums[0];
        for(int i = 1; i < nums.length; i++){
            int num = nums[i];
            int tempMin = Math.min(minSoFar * num, maxSoFar * num);
            int tempMax = Math.max(minSoFar * num, maxSoFar * num);
            minSoFar = Math.min(num, tempMin);
            maxSoFar = Math.max(num, tempMax);
            max = Math.max(maxSoFar, max);
        }
        
        return max;
    }
}

File: 0152. Maximum Product Subarray/dp-kadane.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        (gMin, gMax) = (nums[0], nums[0])
        res = nums[0];
        
        for i in range(1, len(nums)):
            num = nums[i]
            (gMin, gMax) = (
                min(num, gMin * num, gMax * num),
                max(num, gMax * num, gMin * num)
            )
            res = max(res, gMax)
            
        return res;
